#  Cервис для чтения книг по подписке

Необходимо проанализировать базу данных крупного сервиса для чтения книг по подписке.<br>
Цель - сформулировать ценностное предложение для нового продукта.<br>

В базе данных — информация о книгах, издательствах, авторах, а также пользовательские
обзоры книг.

**Описание данных**

Таблица `books`<br>
Содержит данные о книгах:<br>
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Таблица `authors`<br>
Содержит данные об авторах:<br>
- `author_id` — идентификатор автора;
- `author` — имя автора.

Таблица `publishers`<br>
Содержит данные об издательствах:<br>
-`publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

Таблица ratings<br>
Содержит данные о пользовательских оценках книг:<br>
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Таблица reviews<br>
Содержит данные о пользовательских обзорах на книги:<br>
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Загрузка данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine 

In [2]:
# устанавливаем параметры
db_config = {'user': '**********', # имя пользователя
    'pwd': '**********', # пароль
    'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
    'port': 6432, # порт подключения
    'db': 'data-analyst-final-project-db'} # название базы данных

In [3]:
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db']) 

In [4]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [5]:
query = ''' SELECT *
            FROM books 
        '''

In [6]:
books = pd.io.sql.read_sql(query, con = engine)

In [7]:
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [8]:
list = ['books', 'authors', 'publishers', 'ratings', 'reviews']

In [9]:
for i in list:
    query = ''' SELECT *
            FROM {} 
        '''.format(i)
    print(f'Таблица {i}')
    i = pd.io.sql.read_sql(query, con = engine)
    display(i.head())
    i.info()
    print()
    print()

Таблица books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


Таблица authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


Таблица publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


Таблица ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


Таблица reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB




Данные 5 таблиц успешно загружены. Во всех таблицах отсутствуют пропуски. Даты в таблице `books` потребуют перевода в более корректный для дат тип данных.

## Количество книг вышедших после 1 января 2000 года

Посчитаем, сколько книг вышло после 1 января 2000 года.

In [10]:
query = ''' SELECT COUNT(book_id) as books_count
            FROM books
            WHERE CAST(publication_date as timestamp) > '2000-01-01'
        '''
books_count = pd.io.sql.read_sql(query, con = engine)
books_count

books_count
0          819

In [11]:
print(f"Количесто книг, вышедших после 1 января 2020 - {books_count['books_count'][0]}")

Количесто книг, вышедших после 1 января 2020 - 819


## Количество обзоров и средняя оценка для каждой книги

Посчитаем для каждой книги количество обзоров и среднюю оценку.

In [12]:
query = ''' SELECT 
                b.book_id, 
                title, 
                COUNT(DISTINCT review_id) as reviews_count, 
                ROUND(AVG(rating), 1) as raiting_avg
            FROM books as b
            LEFT JOIN ratings as r on b.book_id = r.book_id
            LEFT JOIN reviews as rev on b.book_id = rev.book_id
            GROUP BY b.book_id
            ORDER BY book_id
            
        '''
books_ratings_reviews = pd.io.sql.read_sql(query, con = engine)
books_ratings_reviews.head()

book_id                                              title  reviews_count  \
0        1                                       'Salem's Lot              2   
1        2                 1 000 Places to See Before You Die              1   
2        3  13 Little Blue Envelopes (Little Blue Envelope...              3   
3        4  1491: New Revelations of the Americas Before C...              2   
4        5                                               1776              4   

   raiting_avg  
0          3.7  
1          2.5  
2          4.7  
3          4.5  
4          4.0

Для каждой книги посчитано количество обзоров и средний рейтинг.

## Издательство, которое выпустило наибольшее число книг толще 50 страниц

Определим издательство, которое выпустило наибольшее число книг толще 50 страниц, чтобы исключить из анализа брошюры.

In [13]:
query = ''' SELECT p.publisher_id, p.publisher, COUNT(book_id)
            FROM  publishers as p
            LEFT JOIN books as b on b.publisher_id = p.publisher_id
            WHERE num_pages > 50
            GROUP BY p.publisher_id, p.publisher
            ORDER BY count DESC
            LIMIT 1
           
        '''
books_publishers = pd.io.sql.read_sql(query, con = engine)
books_publishers.head()

publisher_id      publisher  count
0           212  Penguin Books     42

Наибольшее число книг толще 50 страниц выпустило издательство Penguin Books. Число таких книг составило - 42.

## Автор с самой высокой средней оценкой книг для книг с 50 и более оценками

Определим автора с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками.

In [14]:
# итоговая исправленная версия
query = ''' SELECT author, ROUND(AVG(rating), 3) as rating_avg
            FROM  books as b
            LEFT JOIN authors as a on b.author_id = a.author_id
            LEFT JOIN ratings as r on b.book_id = r.book_id
            WHERE r.book_id IN (
                SELECT b.book_id
                FROM  books as b
                LEFT JOIN authors as a on b.author_id = a.author_id
                LEFT JOIN ratings as r on b.book_id = r.book_id
                GROUP BY b.book_id
                HAVING COUNT(DISTINCT rating_id) >= 50)
            GROUP BY author
            ORDER BY rating_avg DESC
            LIMIT 1

        '''
books_publishers = pd.io.sql.read_sql(query, con = engine)
books_publishers.head(10)

author  rating_avg
0  J.K. Rowling/Mary GrandPré       4.287

J.K. Rowling - автор с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками.

## Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок

Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [15]:
query = '''     
            SELECT ROUND(AVG(t.review_count), 2) as reviews_avg
            FROM
                (SELECT 
                    rev.username, 
                    COUNT(DISTINCT r.rating_id) as rating_count, 
                    COUNT(DISTINCT rev.review_id) as review_count
                FROM reviews as rev
                LEFT JOIN ratings as r on rev.username = r.username
                GROUP BY rev.username 
                HAVING COUNT(DISTINCT r.rating_id) > 50) as t



        '''
reviews_books_ratings = pd.io.sql.read_sql(query, con = engine)
reviews_books_ratings

reviews_avg
0        24.33

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.33

Таким образом:
- в сервисе присутствует 819 книг, вышедших после 1 января 2020;
- наибольшее число книг толще 50 страниц (т.е. не брошюры) выпустило издательство Penguin Books, число таких книг составило - 42 книги;
- J.K. Rowling - автор с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками;
- помимо оценок, пользователи пишут обзоры - cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24 обзора. 

Так же для каждой книги посчитано количество обзоров и средний рейтинг.

In [16]:
books_ratings_reviews.head()

book_id                                              title  reviews_count  \
0        1                                       'Salem's Lot              2   
1        2                 1 000 Places to See Before You Die              1   
2        3  13 Little Blue Envelopes (Little Blue Envelope...              3   
3        4  1491: New Revelations of the Americas Before C...              2   
4        5                                               1776              4   

   raiting_avg  
0          3.7  
1          2.5  
2          4.7  
3          4.5  
4          4.0